cleanrl/ppo_continuous_action_envpool_jax.py

```diff
@@ -0,0 +1,382 @@
+        b_obs = obs.reshape((-1,) + envs.single_observation_space.shape)
+        b_logprobs = logprobs.reshape(-1)
+        b_actions = actions.reshape((-1,) + envs.single_action_space.shape)
+        b_advantages = advantages.reshape(-1)
+        b_returns = returns.reshape(-1)
+        values.reshape(-1)
+
+        def ppo_loss(agent_params, x, a, logp, mb_advantages, mb_returns):
+            newlogprob, _, newvalue = get_action_and_value2(x, a, agent_params)
+            logratio = newlogprob - logp
+            ratio = jnp.exp(logratio)
+            approx_kl = ((ratio - 1) - logratio).mean()
+
+            if args.norm_adv:
+                mb_advantages = (mb_advantages - mb_advantages.mean()) / (mb_advantages.std() + 1e-8)
+
+            # Policy loss
+            pg_loss1 = -mb_advantages * ratio
+            pg_loss2 = -mb_advantages * jnp.clip(ratio, 1 - args.clip_coef, 1 + args.clip_coef)
+            pg_loss = jnp.maximum(pg_loss1, pg_loss2).mean()
+
+            # Value loss
+            v_loss = 0.5 * ((newvalue - mb_returns) ** 2).mean()
+
+            # entropy_loss = entropy.mean()
+            # loss = pg_loss - args.ent_coef * entropy_loss + v_loss * args.vf_coef
+            loss = pg_loss + v_loss * args.vf_coef
+            return loss, (pg_loss, v_loss, approx_kl)
+
+        ppo_loss_grad_fn = jax.value_and_grad(ppo_loss, has_aux=True)
+
+        b_inds = jnp.arange(args.batch_size)
+        # clipfracs = []
+        for _ in range(args.update_epochs):
+            key, subkey = jax.random.split(key)
+            b_inds = jax.random.shuffle(subkey, b_inds)
+            for start in range(0, args.batch_size, args.minibatch_size):
+                end = start + args.minibatch_size
+                mb_inds = b_inds[start:end]
+                (loss, (pg_loss, v_loss, approx_kl)), grads = ppo_loss_grad_fn(
+                    agent_params,
+                    b_obs[mb_inds],
+                    b_actions[mb_inds],
+                    b_logprobs[mb_inds],
+                    b_advantages[mb_inds],
+                    b_returns[mb_inds],
+                )
+                updates, agent_optimizer_state = agent_optimizer.update(grads, agent_optimizer_state)
+                agent_params = optax.apply_updates(agent_params, updates)
+
+        return loss, pg_loss, v_loss, approx_kl, key, agent_params, agent_optimizer_state
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs = envs.reset()
+    next_done = np.zeros(args.num_envs)
+    num_updates = args.total_timesteps // args.batch_size
+
+    for update in range(1, num_updates + 1):
+        # Annealing the rate if instructed to do so.
+        if args.anneal_lr:
+            frac = 1.0 - (update - 1.0) / num_updates
+            lrnow = frac * args.learning_rate
+            agent_optimizer_state[1].hyperparams["learning_rate"] = lrnow
+            agent_optimizer.update(agent_params, agent_optimizer_state)
```

Comment:
<img width="1090" alt="image" src="https://user-images.githubusercontent.com/5555347/175840770-a37722cc-c012-4af2-8e52-64bbecbc39c2.png">


It turns out these 4 lines of code slow down the throughput by a half. We are going to need a better learning rate annealing paradigm probably using the official API.

Reply:
From my experience, there's a gain if the main for loop can be replaced with `lax.fori_loop`